### PR TITLE
feat(canvas): add #scrollToElement function

### PR DESCRIPTION
**Pull request by marstamm**
* adjust AutoPlace to scroll to new element
* adjust SearchPad to use centralized function

related to https://github.com/camunda/camunda-modeler/issues/1249
replaces https://github.com/bpmn-io/bpmn-js/pull/1437
recording (3)

### CODE REVIEW

**Comment by philippfromme** [comment]
Hint: It would have been nice to separate this PR into two commits, one that adds the new canvas method and one that adds the new scroll behavior. 😉

**Comment by marstamm** [comment]
That makes sense, I followed the "one commit per PR" rule to closely :smile:

**Comment by MaxTru** [comment]
> That makes sense, I followed the "one commit per PR" rule to closely 
> 
> EDIT: we do not have this rule, I just misunderstood the definition of done :D

When I was onboarded I had the same confusion because of the sentence `a single commit closes the issue via Closes #issuenr` in our https://github.com/bpmn-io/internal-docs/blob/master/development/DEFINITION_OF_DONE.md#definition-of-done.

Since we both had that confusion I propose to re-formulate it to say
`the last commit in the PR closes the issue via Closes #issuenr`

I will create a PR

**Comment by nikku** [comment]
Does not seem to work great when zoomed in or out yet. Also notice how in the search case you may want to provide a larger top padding to not hide behind the search widget / dropdown.

capture pOkQwz_optimized

**Comment by marstamm** [comment]
Thank you for the hints :+1: 

I made the scrolling zoom-aware and added 400px top-padding when used in the search box

**Review by nikku — CHANGES_REQUESTED** [review]
One last remark (and I hope we'll be done after this):

As you can see in the following screen capture the functionality is not stable (indempotent, whatever). If an element scrolled into origin once I'd assume further scrolling invocations to not scroll somewhere else.

This is an issue in the search case. Is it possible to address this?

capture KjwMvh_optimized

**Comment by marstamm** [comment]
What would be the expected behavior when the element is to big to fit into view (with padding)? I would focus on the middle-top, as less get's hidden by UI elements and Sub-Process have the label there

**Comment by philippfromme** [comment]
> What would be the expected behavior when the element is to big to fit into view (with padding)?

I'd say, either center-center or top-left.

**Comment by marstamm** [comment]
These are our options:
**Center-Center** 
note that we center in the padded viewbox, so it looks off-center.
recording (4)

**top-left** 
recording (5)

**Top-Center**
recording (6)

IMO, top-left looks better for Participants and Top-Center for Sub-processes. What do you prefer?

**Comment by nikku** [comment]
We need to keep in mind that this is not bpmn-js specific. Therefore I'd ignore the participant case and vote for top-center.

**Comment by philippfromme** [comment]
I'd go for the very generic top-left, which always gives you a corner.

**Comment by marstamm** [comment]
+1 for Philipp. This also is the best when focusing on Flows while modeling left-right. For example in this model, every other method would focus on a point outside of the path:
image

File: lib/core/Canvas.js [hunk 1]
```diff
@@ -859,6 +862,66 @@ Canvas.prototype.scroll = function(delta) {
...
+  }
+
+  padding = {
+    top: padding.top || defaultPadding,
+    right: padding.right || defaultPadding,
+    bottom: padding.bottom || defaultPadding,
+    left: padding.left || defaultPadding
+  };
+
+  var elementBb = getBoundingBox(element),
```
Comment (philippfromme) [inline lib/core/Canvas.js:889]: Can we rename `elementBb` and other variables that represent bounding boxes to `elementBounds` and so on? We usually use that name. There's also a function [LayoutUtil#asBounds](https://github.com/bpmn-io/diagram-js/blob/develop/lib/layout/LayoutUtil.js#L57) that has this name.

File: lib/core/Canvas.js [hunk 2]
```diff
@@ -859,6 +862,66 @@ Canvas.prototype.scroll = function(delta) {
...
+      elementTrbl = asTRBL(elementBb),
+      viewboxBb = this.viewbox(),
+      zoom = this.zoom(),
+      dx, dy;
+
+  // Shrink viewboxBb with padding
+  viewboxBb.y += padding.top/zoom;
+  viewboxBb.x += padding.left/zoom;
+  viewboxBb.width -= (padding.right + padding.left)/zoom;
```
Comment (philippfromme) [inline lib/core/Canvas.js:898]: Code style: Please leave a space before and after `/`.

File: test/spec/core/CanvasSpec.js [hunk 3]
```diff
@@ -951,6 +951,126 @@ describe('Canvas', function() {
...
+    beforeEach(function() {
+      container = TestContainer.get(this);
+    });
+    beforeEach(createDiagram({ canvas: { width: 500, height: 500 } }));
+
+
+    it('scrolls element into view', inject(function(canvas) {
+
+      // given
+      var element = { x: 650, y: 650, width: 50, height: 50 };
```
Comment (philippfromme) [inline test/spec/core/CanvasSpec.js:965]: I don't like that we're testing with fake elements when the API expects an element? Or does it not? If it accepts bounds then let's make that clear.

**Review by philippfromme — CHANGES_REQUESTED** [review]
Overall very good with nice test coverage. I've got a few remarks.

**Review by nikku — APPROVED** [review]
Looks good from my perspective.

File: test/spec/core/CanvasSpec.js (hunk 3 above)
Comment (marstamm) [reply test/spec/core/CanvasSpec.js]: You are right, I now use real shapes instead of a fake element :+1:

**Review by philippfromme — APPROVED** [review]
👏🏻

**Pull request merged**